src/mcpwm/timer.rs

```diff
@@ -0,0 +1,302 @@
+use core::convert::TryInto;
+use core::marker::PhantomData;
+use core::ptr;
+
+use esp_idf_sys::{
+    esp, mcpwm_del_timer, mcpwm_new_timer, mcpwm_timer_config_t,
+    mcpwm_timer_config_t__bindgen_ty_1, mcpwm_timer_count_mode_t,
+    mcpwm_timer_count_mode_t_MCPWM_TIMER_COUNT_MODE_DOWN,
+    mcpwm_timer_count_mode_t_MCPWM_TIMER_COUNT_MODE_PAUSE,
+    mcpwm_timer_count_mode_t_MCPWM_TIMER_COUNT_MODE_UP,
+    mcpwm_timer_count_mode_t_MCPWM_TIMER_COUNT_MODE_UP_DOWN, mcpwm_timer_enable,
+    mcpwm_timer_handle_t, mcpwm_timer_start_stop,
+    mcpwm_timer_start_stop_cmd_t_MCPWM_TIMER_START_NO_STOP,
+    soc_periph_mcpwm_timer_clk_src_t_MCPWM_TIMER_CLK_SRC_DEFAULT,
+};
+
+use crate::mcpwm::Group;
+use crate::prelude::{FromValueType, Hertz};
+
+use super::operator::NoOperator;
+use super::timer_connection::TimerConnection;
+
+#[derive(Clone, Copy, Debug, PartialEq, Eq)]
+pub struct TimerConfig {
+    resolution: Hertz,
+    period_ticks: u16,
+    count_mode: CountMode,
+    // TODO
+    // on_full: FF,
+    // on_empty: FE,
+    // on_stop: FS,
+}
+
+impl Default for TimerConfig {
+    fn default() -> Self {
+        Self {
+            resolution: 80.MHz().into(),
+            period_ticks: 8_000, // 10kHz
+            count_mode: CountMode::Up,
+        }
+    }
+}
+
+impl TimerConfig {
+    /*#[must_use]
+    pub fn resolution(mut self, resolution: impl Into<Hertz>) -> Self {
+        self.resolution = resolution.into();
+        self
+    }*/
+
+    // TODO: make sure this description is accurate
+    /// Set number of ticks per period
+    ///
+    /// This is inversely proportional to the frequency of the signal
+    ///
+    /// You can calculate the frequency as
+    /// `frequency = resolution / period_ticks`
+    ///
+    /// For example a resolution of 80MHz and a period_ticks of 8_000:
+    /// `10kHz = 80MHz / 8_000`
+    ///
+    /// NOTE: This will be the same as `Self::get_period_peak` for all `CounterMode` except for
+    /// `CounterMode::UpDown` where the period will be twice as large as the peak value since
+    /// the timer will count from zero to peak and then down to zero again
+    #[must_use]
+    pub fn period_ticks(mut self, period_ticks: u16) -> Self {
+        self.period_ticks = period_ticks;
+        self
+    }
+
+    #[must_use]
+    pub fn count_mode(mut self, counter_mode: CountMode) -> Self {
+        self.count_mode = counter_mode;
+        self
+    }
+}
+
+pub struct Timer<const N: u8, G: Group> {
+    _group: G,
+    handle: mcpwm_timer_handle_t,
+    _timer: TIMER<N, G>,
+    /// Number of ticks within a period
+    ///
+    /// See `Self::get_period_ticks` for more info
+    period_ticks: u32,
+
+    /// This is the maximum value that the comparator will see
+    ///
+    /// See `Self::get_period_peak` for more info
+    period_peak: u16,
+}
+
+impl<const N: u8, G: Group> Timer<N, G> {
+    pub fn new(timer: TIMER<N, G>, config: TimerConfig) -> Self {
```

Comment:
The `timer` peripheral should be passed as (roughly) `timer: impl Peripheral<P = impl Timer> + 'd` where `Timer` (to be renamed to `TimerDriver`) is parameterized by the `'d` lifetime. The current implementation does not allow for taking a `&mut ref` to the timer peripheral, which means that dropping the driver will always lose the peripheral.